### PR TITLE
[RFC] Inccomand substitution: Fix search highlight

### PR DESCRIPTION
**Pull request by KillTheMule**
Deactivate search highlighting dring incsub, and replace it by our own. 

Based on https://github.com/neovim/neovim/pull/7231.

Closes https://github.com/neovim/neovim/issues/5598.

### CODE REVIEW

File: src/nvim/ex_cmds.c [hunk 1]
```diff
@@ -6156,7 +6249,11 @@ void ex_substitute(exarg_T *eap)
   curwin->w_p_cul = false;    // Disable 'cursorline'
   curwin->w_p_cuc = false;    // Disable 'cursorcolumn'
 
+  long numval = 0;
+  get_option_value((char_u *)"hls", &numval, NULL, 0);
+  set_option_value((const char *)"hls", 0, NULL, OPT_GLOBAL);
   buf_T *preview_buf = do_sub(eap, profile_setlimit(p_rdt));
+  set_option_value((const char *)"hls", numval, NULL, OPT_GLOBAL);
```
Comment (justinmk) [inline src/nvim/ex_cmds.c]: If possible, it's often faster to set/reset options directly (e.g. `p_hls` in this case), particuarly if there's side-effects. I usually look around the codebase to see how it's typically done for a given option (didn't check `p_hls`), some options will behave badly if their side-effects aren't taken care of. 

In this case I would _guess_ that `p_hls = false` is enough to get what we want, without any problems. Maybe you tried it already though.

Comment (KillTheMule) [reply src/nvim/ex_cmds.c]: I did neither try nor think of it :) I'll give it a a shot, but serious consideration will have to wait until the weekend.

**Comment by KillTheMule** [comment]
I've changed setting `p_hls` to a direct call, and as far as I can see, that works, too. I've also added a little more testing, to show off how it's supposed to work :)

**Comment by justinmk** [comment]
@KillTheMule would you mind rebasing? I think we should just merge this.

**Comment by KillTheMule** [comment]
Done.

**Comment by KillTheMule** [comment]
There seems to be a failure in the oldtests on Travis, which did not happen before the rebase. I'll investigate tomorrow.

**Comment by justinmk** [comment]
@KillTheMule that build failure is on master, don't worry about it.

File: src/nvim/ex_cmds.c [hunk 2]
```diff
@@ -3999,22 +4050,30 @@ static buf_T *do_sub(exarg_T *eap, proftime_T timeout)
...
       set_string_option_direct((char_u *)"icm", -1, (char_u *)"", OPT_FREE,
                                SID_NONE);
-    } else if (*p_icm != NUL && matched_lines.size != 0 && pat != NULL) {
+    } else if (*p_icm != NUL &&  pat != NULL) {
+      if (pre_src_id == 0) {
+        // Get a unique new src_id, saved in a static
+        pre_src_id = bufhl_add_hl(NULL, 0, -1, 0, 0, 0);
+      }
+      if (pre_hl_id == 0) {
+        pre_hl_id = syn_check_group((char_u *)"Substitute", 13);
```
Comment (justinmk) [inline src/nvim/ex_cmds.c]: This should be `10`, right? I'll change it to `sizeof("Substitute") - 1`.

**Comment by justinmk** [comment]
@KillTheMule Nice work!

File: src/nvim/ex_cmds.c (hunk 2 above)
Comment (KillTheMule) [reply src/nvim/ex_cmds.c]: Ah yeah, thats a leftover from "IncSubstitute" I guess :)

**Pull request merged**